contributors/design-proposals/service_links.md

```diff
@@ -0,0 +1,237 @@
+to inject the value of the default variable.
+
+Explicit links offer an opportunity to ease the experience and make custom
+variables easier to reason about.  Instead of referencing variables that will be
+present at runtime, but are not present otherwise in the pod spec, a pod author
+could simply specify a format or prefix for the variables to be created.
+
+## Analysis
+
+Determining the right API is a complex topic that warrants some analysis.
+
+### What degree of specificity is desired?
+
+The desired degree of specificity for an explicit service link is important to
+consider.  There is a continuum of options of different complexities.  On the
+simple end of the spectrum is an API that manifests environment variables in the
+exact same way as the current implicit mechanism, but allows a service to be
+explicitly referenced.  On the complex end of the spectrum is an API that allows
+individual aspects of a service to be projected into individual environment
+variables.
+
+The right API is probably not either of those options, but it is useful to
+consider the pros and cons of both as a way to discover the right API.
+
+#### Pro/con: Simple API
+
+Pros:
+
+1.  Very small practical and conceptual delta from current behavior
+2.  Users can employ the same strategies for custom env vars
+3.  Easiest to explain
+4.  Easiest to support
+
+Cons:
+
+1.  Inflexible
+2.  Users have the same usability issues when custom env var names are required
+3.  Does not extend well to services in other namespaces
+
+#### Pro/con: Most complex API
+
+Pros:
+
+1.  Offers highest degree of flexibility
+2.  Custom env var names are cheap at fine-grained level of control
+
+Cons:
+
+1.  Most difficult to explain and reason about
+2.  Most difficult to support
+3.  Higher implementation cost
+4.  Large conceptual delta
+5.  Likely to add significant complexity to pod spec
+6.  Likely to be very verbose
+
+#### Lessons Learned from Prior APIs
+
+One fundamental lesson of Kubernetes API design is that we must make backward
+compatible API changes except in cases of special exception.  This means that we
+need to consider backward compatibility when making new API changes.  As a
+requirement, backward compatibility imposes a burden that you have to live with
+the APIs you release.  This means that we should attempt to be conservative with
+new API features and avoid unnecessary complexity.  Therefore, a simpler API is
+advantageous because complexity can be added later **if desired and warranted**.
+
+#### Middleground
+
+A conservative middleground might be an API that has behavior extremely close to
+the current implicit behavior but allows specifying custom variable names.  This
+approach has the advantage of being close to what users already expect, and can
+be implemented with a fairly simple API.
+
+### Non-existent Services
+
+When a service referenced using this API does not exist, the Kubelet should
+avoid starting the pod.
```

Comment:
Kubelet should not reject it. The same should be true for secrets, configmaps, PVCs, and anything else a pod references.

Ideally, the workload controller (e.g., RC, RS, DS, SS) should not create pods until dependencies exist. However, barring that Kubelet should wait until the dependencies exist. I'd also like to be able to create controllers and even pods without any specified image, with the desired image provided later.

Today, kubectl creates resources in a directory in an arbitrary order and even concurrently. I'd like to preserve and strengthen the ability to do that, as well as for asynchronous automation to provide additional resources and/or fields, potentially even without initializers.

As for services without endpoints, not sure we want to go there yet.